File: apina-core/src/main/java/fi/evident/apina/utils/PropertyUtils.java
package fi.evident.apina.utils;

import static fi.evident.apina.utils.StringUtils.uncapitalize;

public final class PropertyUtils {

    private PropertyUtils() { }

    public static String propertyNameForGetter(String getterName) {
        if (getterName.startsWith("get"))
            return uncapitalize(getterName.substring(3));
        else if (getterName.startsWith("is"))
            return uncapitalize(getterName.substring(2));
        else
            throw new IllegalArgumentException("not a valid name for getter "+ getterName);
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/utils/PropertyUtils.kt
package fi.evident.apina.utils

fun propertyNameForGetter(getterName: String): String =
    // Value/inline classes result in name suffixed with hash
    propertyNameForSimpleGetter(getterName.substringBefore('-'))

private fun propertyNameForSimpleGetter(getterName: String): String =
    when {
        getterName.startsWith("get") -> getterName.removePrefix("get").replaceFirstChar { it.lowercase() }
        getterName.startsWith("is") -> getterName.removePrefix("is").replaceFirstChar { it.lowercase() }
        else -> error("not a valid name for getter $getterName")
    }
